runtime-testsuite/src/org/antlr/v4/testgen/TestGenerator.java

```diff
@@ -30,202 +30,315 @@
-					new File(templatesDir),
-					new File(targetTemplate),
-					viz);
-		gen.info("Generating target " + gen.getTargetName());
-		gen.execute();
+        try {
+            ListIterator<String> moreArgs = Arrays.asList(args).listIterator(0);
+            while (moreArgs.hasNext()) {
+                String arg = moreArgs.next();
+                if (!gen.acceptOption(arg, moreArgs))
+                    throw new TestGenException("Option not recognized: %s", arg);
+            }
+
+            gen.applyDefaults();
+            gen.showOptions();
+	        gen.execute();
+
+        } catch (TestGenException ex) {
+            gen.errPrintln(ex.getClass().getSimpleName() + ": " + ex.getMessage());
+            for (StackTraceElement ste : ex.getStackTrace())
+                gen.errPrintln("\tat " + ste);
+            System.exit(1);
+        }
 	}
 
+    public TestGenerator() {
+    }
+
+    // set by acceptOption() and applyDefaults() ...
+    public File rootDir;
+    public File outDir;
+    public File testTemplatesRoot;
+    public String encoding = "UTF-8";
+    public boolean visualize;
+    public boolean browsers;
+    public Collection<String> targets = new TreeSet<String>(String.CASE_INSENSITIVE_ORDER);
+    public Collection<String> excludes = new TreeSet<String>();
+
+    // set during genTarget()
+    protected String targetName;
+    protected String lineSeparator = System.getProperty("line.separator");
+    protected ErrorBuffer errorBuffer = new ErrorBuffer();
+
 	// This project uses UTF-8, but the plugin might be used in another project
 	// which is not. Always load templates with UTF-8, but write using the
 	// specified encoding.
-	protected final String encoding;
-	protected final String targetName;
-	protected final String rootDir;
-	protected final File outputDir;
-	protected final File testTemplates;
-	protected final File runtimeTemplate;
-	protected final boolean visualize;
-
-	public TestGenerator(String encoding, String targetName, String rootDir, File outputDir, File testTemplates, File runtimeTemplate, boolean visualize) {
-		this.encoding = encoding;
-		this.targetName = targetName;
-		this.rootDir = rootDir;
-		this.outputDir = outputDir;
-		this.testTemplates = testTemplates;
-		this.runtimeTemplate = runtimeTemplate;
-		this.visualize = visualize;
-	}
 
-	private String getTargetName() {
-		return targetName;
-	}
+    /**
+     * Consume a command line option.
+     * 
+     * @param arg       the argument to examine
+     * @param moreArgs  The rest of the arguments.  If arg is recognized as an
+     *                  option that requires a value, the value is consumed from
+     *                  this iterator; TestGenException is thrown if not enough.
+     * @return true if option was recognized and consumed; false if unrecognized
+     * @throws TestGenException
+     */
+    public boolean acceptOption(String arg, Iterator<String> moreArgs) {
+        try {
+            if ("-root".equals(arg))
```

Comment:
:bulb: Several control flow blocks in new code are missing braces. Obviously this was also the case for the original code. It would help if you could add them to the new code, otherwise I can submit a pull request in the future to address it separately. (I wasn't going to comment, but this is a long control flow block so it really stood out to me.)